### PR TITLE
Factory boy

### PR DESCRIPTION
**Pull request by WKD622**
Faster than I expected I managed to rewrite my code to use factory_boy. Proud to say that it's ready to be checked.

### CODE REVIEW

File: api/factories.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+class EventFactory(factory.Factory):
+    class Meta:
+        model = Event
+
+    name = factory.Sequence(lambda n: 'event ' + str(n))
+    description = factory.sequence(lambda n: 'description ' + str(n))
+
+
+class PastEventFactory(EventFactory):
+    datetime = factory.fuzzy.FuzzyDateTime(datetime.datetime(2011, 8, 15, 8, 15, 12, 0, pytz.UTC),
```
Comment (mciszczon) [inline api/factories.py]: Why do you use this FuzzyDateTime here? And also, shouldn't it be replace with Faker (as noted in the Factory Boy documentation)?

Comment (mciszczon) [reply api/factories.py]: I would also change the date to something like `datetime.today() - datetime.timedelta(days=x)` to keep this hardcoded date off from here.

File: manager/settings.py [hunk 2]
```diff
@@ -113,7 +113,7 @@
 
 USE_L10N = True
 
-USE_TZ = False
+USE_TZ = True
```
Comment (mciszczon) [inline manager/settings.py:116]: If you want to use time zones, then you should use Django's own `timezone` module instead of the standard `datetime`.

`from django.utils import timezone` (it has got all the same members as the standard `datetime`).

**Review by mciszczon — APPROVED** [review]

File: api/factories.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import pytz
+from .models import *
+
+
+class EventFactory(factory.Factory):
+    class Meta:
+        model = Event
+
+    name = factory.Sequence(lambda n: 'event ' + str(n))
+    description = factory.sequence(lambda n: 'description ' + str(n))
```
Comment (muchas) [inline api/factories.py]: `factory.sequence` -> `factory.Sequence` (typo)

File: api/factories.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class PastEventFactory(EventFactory):
+    datetime = factory.fuzzy.FuzzyDateTime(datetime.datetime(2011, 8, 15, 8, 15, 12, 0, pytz.UTC),
+                                           (datetime.datetime.today() - datetime.timedelta(days=1)).replace(
+                                               tzinfo=timezone.utc))
+
+
+class FutureEventFactory(EventFactory):
```
Comment (muchas) [inline api/factories.py:29]: I'm not sure if having classes like `PastEventFactory`, `TodayEventsFactory` and `FutureEventFactory` are necessary, my thoughts are:
- those classes differ by one field only (datetime), it may be easier to use `EventFactory` with parameter,
- when it comes to testing dates and time it's usually good idea to test edge cases with extreme values where small time change may cause behaviour change, e.g
```
past = timezone.now() - datetime.timedelta(seconds=1)
past_event = EventFactory(datetime=past)
```

What do you think?

File: api/tests/test_models.py [hunk 5]
```diff
@@ -1,40 +1,32 @@
...
+    past_event = PastEventFactory.create()
     past_event.save()
 
-    future_events = [
-        Event(name='name2', description='description2', datetime=today + timezone.timedelta(days=1)),
-        Event(name='name2', description='description2', datetime=today + timezone.timedelta(days=30)),
-        Event(name='name3', description='description3', datetime=today + timezone.timedelta(days=360)),
-    ]
+    future_events = FutureEventFactory.create_batch(3)
     for event in future_events:
```
Comment (muchas) [inline api/tests/test_models.py]: AFAIK this for loop can be removed too. `FutureEventFactory.create_batch(3)` already persists model in the database.

File: api/tests/test_models.py [hunk 6]
```diff
@@ -1,40 +1,32 @@
...
 
 
 @pytest.mark.django_db
 def test_future_events():
     # given
-    today = timezone.now()
-
-    past_event = Event(name='name1', description='description1', datetime=today - timezone.timedelta(days=1))
+    past_event = PastEventFactory.create()
     past_event.save()
```
Comment (muchas) [inline api/tests/test_models.py]: I think this line can be removed. `PastEventFactory.create()` should persist object in the database.

File: api/tests/test_models.py [hunk 7]
```diff
@@ -1,40 +1,32 @@
...
-
-    past_event = Event(name='name1', description='description1', datetime=today - timezone.timedelta(days=3))
-    todays_event = Event(name='name2', description='description2', datetime=today)
-    future_event = Event(name='name3', description='description3', datetime=today + timezone.timedelta(days=2))
+    past_event = PastEventFactory.create()
+    todays_event = TodayEventsFactory.create()
+    future_event = FutureEventFactory.create()
     past_event.save()
     todays_event.save()
     future_event.save()
```
Comment (muchas) [inline api/tests/test_models.py]: Same as above - those three lines can go away :)

File: api/tests/test_models.py [hunk 8]
```diff
@@ -1,40 +1,32 @@
...
 from ..models import Event
 
 
 @pytest.mark.django_db
 def test_future_events():
     # given
-    today = timezone.now()
-
-    past_event = Event(name='name1', description='description1', datetime=today - timezone.timedelta(days=1))
+    past_event = PastEventFactory.create()
```
Comment (muchas) [inline api/tests/test_models.py]: FYI: this call is the same as `past_event = PastEventFactory()`. The latter is used in the codebase more often.

**Review by muchas — CHANGES_REQUESTED** [review]
Well done, few minor comments, but overall you've quickly harnessed `factory_boy` 👍

File: api/tests/test_models.py (hunk 5 above)
Comment (WKD622) [reply api/tests/test_models.py]: I checked and it doesn't work without saving.

File: api/tests/test_models.py (hunk 6 above)
Comment (WKD622) [reply api/tests/test_models.py]: same as above

File: api/tests/test_models.py (hunk 7 above)
Comment (WKD622) [reply api/tests/test_models.py]: same as above

File: api/tests/test_models.py (hunk 8 above)
Comment (WKD622) [reply api/tests/test_models.py]: In my case tests do not pass without use of create() function.

File: api/factories.py (hunk 3 above)
Comment (WKD622) [reply api/factories.py]: done

File: api/factories.py (hunk 4 above)
Comment (WKD622) [reply api/factories.py:29]: Okey. I wanted to check inheritance in factory_boy, only to learn some more things. I agree that it is kind of a overkill but now i know how to do it :).

When it comes to dates - okey. I will correct them

File: manager/settings.py (hunk 2 above)
Comment (WKD622) [reply manager/settings.py:116]: ok

File: api/tests/test_models.py (hunk 5 above)
Comment (muchas) [reply api/tests/test_models.py]: I think we've missing one thing: https://factoryboy.readthedocs.io/en/latest/orms.html#the-djangomodelfactory-subclass

`EventFactory` should be a subclass of `DjangoModelFactory` to use that feature.
